ImageData/CartaHdf5Image.cc

```diff
@@ -136,36 +137,79 @@ casacore::Bool CartaHdf5Image::doGetMaskSlice(casacore::Array<bool>& buffer, con
-        SetupImageInfo(attributes);
-    } catch (casacore::HDF5Error& err) {
-        hdf5_group.close();
-        hdf5_file.close();
+        casacore::Record header = ConvertInfoToCasacoreRecord(info);
+        valid = SetupCoordSys(header);
+        SetupImageInfo(header);
+    } catch (casacore::AipsError& err) {
+        std::cerr << "Error opening HDF5 image: " << err.getMesg() << std::endl;
     }
     return valid;
 }
 
-bool CartaHdf5Image::SetupCoordSys(casacore::Record& attributes) {
-    // Use header attributes to set up Image CoordinateSystem
-    if (attributes.empty())
+casacore::Record CartaHdf5Image::ConvertInfoToCasacoreRecord(const CARTA::FileInfoExtended* info) {
+    // convert header_entries to Record
+    casacore::Record header_record;
+    for (int i = 0; i < info->header_entries_size(); ++i) {
+        const CARTA::HeaderEntry header_entry = info->header_entries(i);
+        const std::string entry_name = header_entry.name();
+	if (entry_name == "SIMPLE") {
+            continue; // added in FITS util
+        }
+        switch (header_entry.entry_type()) {
+            case CARTA::EntryType::STRING: {
+                if ((entry_name == "EXTEND") || (entry_name == "BLOCKED")) { // bool
+                    header_record.define(entry_name, (header_entry.value() == "T" ? true : false));
+		} else if ((entry_name == "BITPIX") || (entry_name == "NAXIS")) { // int
+                    header_record.define(entry_name, std::stoi(header_entry.value()));
+		} else if (ShouldBeFloat(entry_name)) { // float
+                    header_record.define(entry_name, std::stof(header_entry.value()));
+                } else { // string
+                    casacore::String entry_value = header_entry.value();
+		    entry_value = (entry_value == "Kelvin" ? "K" : entry_value);
+		    if (entry_name == "DATE_OBS") { // date
+			casacore::String fits_date;
+                        casacore::FITSDateUtil::convertDateString(fits_date, entry_value);
+                        header_record.define(entry_name, fits_date);
+                    } else {
+                        header_record.define(entry_name, entry_value);
+                    }
+                }
+                break;
+            }
+            case CARTA::EntryType::FLOAT:
+                header_record.define(header_entry.name(), static_cast<float>(header_entry.numeric_value()));
+                break;
+            case CARTA::EntryType::INT:
+                header_record.define(header_entry.name(), static_cast<int>(header_entry.numeric_value()));
+                break;
+            default:
+                header_record.define(header_entry.name(), header_entry.value());
+        }
+    }
+    return header_record;
+}
+
+bool CartaHdf5Image::ShouldBeFloat(const std::string& entry_name) {
+    casacore::String name(entry_name);
+    return ((name == "EQUINOX") || (name == "EPOCH") || (name == "LONPOLE") || (name == "LATPOLE") || (name == "RESTFRQ") ||
```

Comment:
Should we maybe have a vector of exact names and a vector of starting substrings and iterate over them? Would that be significantly slower?

Reply:
Will redo this